### ISSUE TITLE
Training time too long

### ISSUE THREAD

**Issue by xwhjy** [daveredrum/Scan2Cap#4]
Thanks for releasing the code! I really appreciate your hard work!

When I followed the instructions and ran the training command `python scripts/train.py --use_multiview --use_normal --use_topdown --use_relation --use_orientation --num_graph_steps 2 --num_locals 10 --batch_size 12 --epoch 50
`, I found that it would take nearly 14 hours (as shown in the screenshot below) to validate simply on the training set. Is it normal to have such a long time period for validation? I use one GeForce RTX 2080 GPU with 11GB of memory.
![Screen Shot 2021-07-27 at 3 19 43 pm](https://user-images.githubusercontent.com/68929040/127100115-ee16e79d-2033-4454-ad29-424de977aed4.jpg)

**Comment by daveredrum** [comment]
Hi @xwhjy, that seems a bit weird to me because the evaluation should only take several minutes. Can you try to disable the evaluation in the first iteration (simply change [this line](https://github.com/daveredrum/Scan2Cap/blob/main/lib/solver.py#L451) to `if (self._global_iter_id+1) % self.val_step == 0:`), and post the screenshot of the solver output for one training iteration?

**Comment by xwhjy** [comment]
Thanks Dave@daveredrum for your reply!

Then it just skipped the evaluation for the first training iteration and continued training without any evaluation all the way to the  2000th iteration. However, It still gets stuck when it enters evaluation mode at `val_step`-th training iteration. 

![Screen Shot 2021-07-28 at 9 24 38 am](https://user-images.githubusercontent.com/68929040/127240053-c6cf0404-8731-4aa3-950b-9e2df7fc10fd.jpg)


I checked the code for evaluation and found that in the model forward part (the main costly part) during evaluation, for each evaluation data (i.e., a caption sample), it generates caption prediction for each candidate object proposal and that's 256 in total. Could this be the reason? Correct me if I'm wrong! Much appreciated!

**Comment by daveredrum** [comment]
I guess the most costly part is loading the data (see `mean_fetch_time`), which should normally take <1s. How many workers did you set for the dataloader? And did you store the ENet multiview features on your local SSD?

**Comment by xwhjy** [comment]
> I guess the most costly part is loading the data (see `mean_fetch_time`), which should normally take <1s. How many workers did you set for the dataloader? And did you store the ENet multiview features on your local SSD?

It is weird. I didn't change the code so used 4 workers to load the data. I also followed the instructions to have the ENet multiview features stored on my local SSD.

Apart from that, as you see the progress bar for the evaluation at iteration 2000, it would take 14 hours. Why it costs so much more time to evaluate than to train?

**Comment by daveredrum** [comment]
![Screenshot from 2021-07-28 13-47-05](https://user-images.githubusercontent.com/22779643/127317185-93feb5a1-22f4-46be-b225-f1210f0015fa.png)

It seems a bit weird to me too - loading the data should take that much time. How many CPUs are you using at the moment?

**Comment by xwhjy** [comment]
There are 16 cpus in the machine. 

Can I also see how the progress bar is showing in your program running?

**Comment by daveredrum** [comment]
eval on train split -> 100%|██████████| 36/36 [14:50<00:00, 24.74s/it]
eval on val split -> 100%|██████████| 9/9 [03:35<00:00, 23.94s/it]

Generating descriptions during inference can for sure be improved, as there are a lot of invalid ones out of the 256 proposals. But it shouldn't take >1 hr to finish the evaluation. Can you try to remove the `--use_multiview` and fetch the data purely from the RAM?

**Comment by xwhjy** [comment]
> eval on train split -> 100%|██████████| 36/36 [14:50<00:00, 24.74s/it]
> eval on val split -> 100%|██████████| 9/9 [03:35<00:00, 23.94s/it]
> 
> Generating descriptions during inference can for sure be improved, as there are a lot of invalid ones out of the 256 proposals. But it shouldn't take >1 hr to finish the evaluation. Can you try to remove the `--use_multiview` and fetch the data purely from the RAM?

I will definitely try that. And I was also wondering if it is because I am using symbolic links to access data. 

For evaluation, in your running, it takes around 24 seconds per iteration. As the train split has 36665 samples and the batch size is 12,  there would be 36665/12=3056 iterations to run. If each iteration takes 24 seconds, then it would be 3056*24/3600=20.3 hours. Similarly, for evaluation on val split, it would take (9508/12)*24/3600=5.2 hours.  It seems in either cases, it would take over 1 hour to finish the evaluation. Please correct me if I am understanding wrong. Really appreciate your time!

**Comment by daveredrum** [comment]
No, the model is evaluated on the scenes NOT on the samples, which means there are only 562 scenes in train and 141 in val to evaluate. The whole evaluation should by all means finish within half an hour.

**Comment by xwhjy** [comment]
> No, the model is evaluated on the scenes NOT on the samples, which means there are only 562 scenes in train and 141 in val to evaluate. The whole evaluation should by all means finish within half an hour.

I am a little bit lost. The val dataset defined here uses ScannetReferenceDataset. https://github.com/daveredrum/Scan2Cap/blob/08410dba79c35bd4edd4a50e8f0c63d0cc6f38a7/scripts/train.py#L287
And as defined here, the length of the dataset is #samples. https://github.com/daveredrum/Scan2Cap/blob/08410dba79c35bd4edd4a50e8f0c63d0cc6f38a7/lib/dataset.py#L318

**Comment by daveredrum** [comment]
The raw data is defined here:

```
        new_scanrefer_eval_train = []
        for scene_id in train_scene_list:
            data = deepcopy(SCANREFER_TRAIN[0])
            data["scene_id"] = scene_id
            new_scanrefer_eval_train.append(data)

        new_scanrefer_eval_val = []
        for scene_id in val_scene_list:
            data = deepcopy(SCANREFER_TRAIN[0])
            data["scene_id"] = scene_id
            new_scanrefer_eval_val.append(data)
```

Only the scene data is loaded - You can check the length of the val dataset.

**Comment by xwhjy** [comment]
Ohh you are right! I finally got it! Sorry for being so dumb these days. Really appreciate your patience! Thanks a lot!

**Comment by yang-Michael** [comment]
@xwhjy Hi, can you tell me how long your training time was after you fixed this bug? And how much space did your cpu memory take up when you started training?
